itou/www/siaes_views/views.py

```diff
@@ -45,57 +51,84 @@ def job_description_card(request, job_description_id, template_name="siaes/job_d
-                    code = job_through.appellation.code
-                    new_custom_name = request.POST.get(f"custom-name-{code}", "")
-                    new_description = request.POST.get(f"description-{code}", "")
-                    new_is_active = bool(request.POST.get(f"is_active-{code}"))
-                    if (
-                        job_through.custom_name != new_custom_name
-                        or job_through.description != new_description
-                        or job_through.is_active != new_is_active
-                    ):
-                        job_through.custom_name = new_custom_name
-                        job_through.description = new_description
-                        job_through.is_active = new_is_active
-                        job_through.save()
-
-            messages.success(request, "Mise à jour effectuée !")
-            return HttpResponseRedirect(reverse_lazy("dashboard:index"))
-
-    context = {"siae": siae}
+        # Validate submitted data: this is not the standard way to do things
+        # and errors will not be shown at the field level.
+        for code in submitted_codes:
+            data = {
+                # Omit `SiaeJobDescription.appellation` since the field is
+                # hidden and `Appellation.objects.get()` will fail anyway.
+                "custom_name": request.POST.get(f"custom-name-{code}", ""),
+                "description": request.POST.get(f"description-{code}", ""),
+                "is_active": bool(request.POST.get(f"is_active-{code}")),
+            }
+            # We use a single ModelForm instance to validate each submitted group of data.
+            form = ValidateSiaeJobDescriptionForm(data=data)
+            if not form.is_valid():
+                for key, value in form.errors.items():
+                    verbose_name = form.fields[key].label
+                    error = value[0]
+                    # The key of the dict is used in tests.
+                    errors[code] = f"{verbose_name} : {error}"
+
+        if not errors:
+
+            current_codes = set(siae.job_description_through.values_list("appellation__code", flat=True))
+
+            codes_to_create = submitted_codes - current_codes
+            # It is assumed that the codes to delete are not submitted (they must
+            # be removed from the DOM via JavaScript). Instead, they are deducted.
+            codes_to_delete = current_codes - submitted_codes
+            codes_to_update = current_codes - codes_to_delete
+
+            if codes_to_create or codes_to_delete or codes_to_update:
+                with transaction.atomic():
+                    # Create.
+                    for code in codes_to_create:
+                        appellation = Appellation.objects.get(code=code)
+                        through_defaults = {
+                            "custom_name": request.POST.get(f"custom-name-{code}", ""),
```

Comment:
Peut être que le form effectue (ou effectuera) des changements sur les data.
J'aurais plutôt collecter le cleaned_data de la validation itérative au lieu d'extraire du POST.

Reply:
Le truc c'est qu'ils ne sortent pas du `cleaned_data` sous ce format `custom-name-{code}`. Je vais voir si j'ai une solution simple.

Reply:
J'ai pas de solution rapide, je reste sur le comportement actuel pour le moment. Le code est assez clair pour faire s'interroger un développeur en cas de changement qui seraient déclenchés par le `form`.